### ISSUE TITLE
[Bug]: Drilldown link urls and repeated variable scope not used

### ISSUE THREAD

**Issue by genkimaru** [grafana/grafana#8424]
firstly , I declare my environment 
-- grafana 4.3.0  belta
-- datasource opentsdb
-- os win10
the confusing things occurs when I using a repeat panel ,  I set a template variable $x , then set the repeat panel  related to the $x . then I set the $x  in title and in url param  respectively.  when I run it ,I can find the title is value I wanted , but the url param isn't same to the title . why ?  the configuration like below
![image](https://cloud.githubusercontent.com/assets/5584476/26244287/b0b8a8d4-3cc1-11e7-81dc-33a0de6461b0.png)

**Comment by torkelo** [comment]
might be  a bug in drilldown link not looking at scoped vars

**Comment by bergquist** [comment]
15:41 < AbbyH> Hi All, I think i've found a bug and the solution for it but i dont have a go environment setup. Would someone be willing to help me test it out?
15:55 < AbbyH> I believe its this issue https://github.com/grafana/grafana/issues/8424 and i think it can be fixed by changing public/app/plugins/panel/singlestat/module.ts:637 the last argument from 
               data.scopedVars to panel.scopedVars
15:57 < faxmodem> AbbyH: add it to the comments
15:58 < faxmodem> AbbyH: and if someone has time to test it out, they'll fix it
15:58 < faxmodem> AbbyH: thanks!
15:58 < AbbyH> just trying to remember by github login to do so :)
16:00 < faxmodem> if you want I can add it and credit you if you give me the name of your github handle
16:07 < AbbyH> give me a few minutes, I need to rename my github profile

**Comment by AbbyRH** [comment]
Back into my account now. Thanks @bergquist

**Comment by bergquist** [comment]
I cannot reproduce this in master

http://play.grafana.org/dashboard/db/prometheus-repeat?refresh=5m&orgId=1&var-node=demo.robustperception.io&var-filesystem_mountpoint=All&var-job=alertmanager&var-job=node

What version are you on?

**Comment by AbbyRH** [comment]
Grafana v4.6.1 (commit: cac8b97)

I think its only on single stat that this is a bug. the link that 'pops up' when you hover over the external link icon is correct, but for single stats it also makes the entire panel a link and that link is is missing the scopedVars

**Comment by bergquist** [comment]
Cant reproduce that either :/ 

http://play.grafana.org/dashboard/db/prometheus-repeat?refresh=5m&orgId=1 (updated dashboard to repeat single stat)

**Comment by AbbyRH** [comment]
looking into the js using chrome's built in debugger it appears the version on the playground has scopedVars set correctly in the data object. On my version data.scopedVars is undefined. 
I'll try and setup a local instance on the same commit as the playground in the morning and confirm if its fixed using my live data

Thanks for your help @bergquist

**Comment by cxcv** [comment]
Hi there,

I can reproduce it; looks like it works for Prometheus, but when Variable Type is Custom (or a MySQL query) it is not using the repeated variable but the whole array.

Cheers
Benjamin

**Comment by craigmorten** [comment]
Reproduced in Grafana v5.2.0-pre1 (8aa05e6):
- Templated variable `x` of `Query` type, using InfluxDB datasource, with `multi-select` and `all` options selected.
- Single Stat panel repeated by the templated variable `x`, with a URL params value of the templated variable `var-x=${x}`.
- Clicking on one of the Single Stat panels (e.g. for ${x} == 1), results in a query param of the form `var-x={1,...}`, whereas expect `var-x=1`

**Comment by torkelo** [comment]
@craigmorten strange, unable to replicate using your steps.

**Comment by craigmorten** [comment]
@torkelo curious... let me try again and will we come back to you with some screenshots and further details - the hardship is I'm currently working with sensitive data so naive screenshots and full detail is out of the question. I'll try to spare some time to set up some dummy instances and get report back later...

**Comment by craigmorten** [comment]
Let me know if this isn't enough of a starting point:

1. Create a new dashboard to contain the single stat, e.g. `Test InfluxDB Drilldown`
1. Configure a `device` variable in the dashboard (we have devices tagged by hostname):
    1. Name: `device`
    1. Label: `device`
    1. Type: `Query`
    1. Datasource: Our influxDB datasource
    1. Query: `SHOW TAG VALUES WITH KEY = hostname`
    1. Regex: using regex to strip out a device number
    1. Mulit-value: ticked
    1. Include All option: ticked

<img width="1115" alt="screen shot 2018-07-02 at 11 30 14" src="https://user-images.githubusercontent.com/33631490/42160099-e93e51fe-7ded-11e8-8756-2acc97b06f12.png">

1. Configure a `device_non_multiselect` variable in the dashboard (this will receive the drilldown variable):
    1. Name: `device_non_multiselect`
    1. Label: `device_non_multiselect`
    1. Type: `Query`
    1. Datasource: Our influxDB datasource
    1. Query: `SHOW TAG VALUES WITH KEY = hostname`
    1. Regex: using regex to strip out a device number
    1. Mulit-value: not ticked
    1. Include All option: not ticked

<img width="1212" alt="screen shot 2018-07-02 at 11 49 15" src="https://user-images.githubusercontent.com/33631490/42160151-12d98f06-7dee-11e8-82c6-7a824e025c11.png">

1. Add a Single Stat Panel to the dashboard.
    1. Title: `${device}`
    1. Repeat:
        1. For each value of `device`
        1. Min Width: 1
        1. Direction: `Horizontal`
    1. Drilldown / detail link:
        1. Type: `dashboard`
        1. Dashboard: `Test InfluxDB Drilldown`
        1. Title: `Test InfluxDB Drilldown`
        1. Url params: `var-device=${device}&var-device_non_multiselect=${device}`
        1. Include time range: ticked
        1. Open in new tab: ticked

<img width="911" alt="screen shot 2018-07-02 at 11 50 41" src="https://user-images.githubusercontent.com/33631490/42160184-31c8d912-7dee-11e8-9fa9-07692e96a9fe.png">

At this point, if I click the Single Stat panel (with tooltip `click to go to - Test InfluxDB Drilldown`), everything is fine, i.e. a new tab opens with the same dashboard, and `device` and `device_non_multiselect` variables are set to the value for the repeated panel that is clicked.

1. Edit Single Stat panel -> Metrics
    1. Delete the `Test data: random walk` query.
    1. Swap `Data Source` to our influxDB datasource from `default`.

<img width="767" alt="screen shot 2018-07-02 at 11 51 23" src="https://user-images.githubusercontent.com/33631490/42160237-58092352-7dee-11e8-8f27-f5b7d9a225dc.png">

Now clicking the Single Stat panel (with tooltip `click to go to - Test InfluxDB Drilldown`) results in `device` and `device_non_multiselect` variables being incorrect. I.e. a new tab opens with the same dashboard and `device` and `device_non_multiselect` variables are set to `{...}`.

<img width="1625" alt="screen shot 2018-07-02 at 11 52 33" src="https://user-images.githubusercontent.com/33631490/42160259-726b99d2-7dee-11e8-8d0b-5c7439b9cc55.png">

My current workaround is to use the equivalent of `var-device=${device:csv}&var-device_non_multiselect=${device:csv}`, though this is not an intuitive, nor likely scalable solution.

**Comment by torkelo** [comment]
So you only get this when you have no metric queries in the metrics tab? I think you need at least one query for this stuff to work.

**Comment by craigmorten** [comment]
@torkelo this occurs with zero, one or more queries in the metrics tab - the thing that seems to break it is the fact that the datasource is set to and InfluxDB one.

**Comment by cxcv** [comment]
Hi @craigmorten @torkelo 

I think the behaviour is independent from InfluxDB; looks like singlestat is using the wrong variable scope for the url parameters. Here's a dashboard that reproduces this with a custom variable:
https://pastebin.com/2cdTWArz (on pastebin as I don't know if/how I can create on play.grafana.org)
- title, arrow top-left and anything looks good - the current repeat value is used
- clicking on the centered text is wrong - the array is used

Can you reproduce it?

Cheers
Benjamin

**Comment by drenalin23** [comment]
I was able to reproduce something similar with a prometheus data source.  Created a repeating singlestat panel, that repeated on a template query variable.  I could create the case where that URL params would appear to get the entire list, but the description link in the upper left was properly created when the 'Format as' was set to Table (which looks like underneath also sets Instant: false, I had to manually remove that as it seems when you switch back to Time-Series, instant:false doesn't get removed from the panel JSON).  It was driving me crazy because this appeared to be working when I created this as a new repeating set (but didn't inadvertently set to Table).  Not sure if that is what happened to folks above, but something to check out.   hth.

**Comment by craigmorten** [comment]
@torkelo finally got round to looking at source code. The issue arises from the logic of where the scopedVars are retrieved.

Here https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/singlestat/module.ts#L655 the `scopedVars` for drilldown links are retrieved from the `data` object.

This property of `data` is added in the `setValues` function here https://github.com/grafana/grafana/blob/master/public/app/plugins/panel/singlestat/module.ts#L323 

Thus, if there are no series / data, the scoping per panel ceases to work. This is undesirable (This may be the cause of some other issues against Grafana atm e.g. https://github.com/grafana/grafana/issues/11441).

Suggested solutions:

1. Move on Line 323:

    ```js
    data.scopedVars = _.extend({}, this.panel.scopedVars);
    ```

    prior to the if statement `if (this.series && this.series.length > 0) {...}`.

2. Or use `panel.scopedVars` on Line 655 as done elsewhere, e.g.

    ```js
    linkInfo = linkSrv.getPanelLinkAnchorInfo(panel.links[0], data.scopedVars);
    ```

This may only fix the singlestat, yet to test out elsewhere. A more global fix may be required.
